OpenSim/Common/Exception.h

```diff
@@ -47,12 +47,55 @@
 //=============================================================================
 //=============================================================================
 
-namespace OpenSim { 
+namespace OpenSim {
+
+
+/** Macros to throw OpenSim exceptions. The purpose of these macros is to avoid
+having to provide the first few arguments (which are common to all OpenSim
+exceptions) to the exception constructor in the throw statements. This also
+allows us to add more details (eg class name) later easily.                   */
+
+// Note -- Extra braces enclosing "if" are to avoid problems when these macros 
+// are called within if-else statements like:
+// \code{.cpp}
+//           if(<some condition>)
+//               OPENSIM_THROW_IF(<arguments>)
+//           else
+//               <some statements>
+// \endcode                                                                     
+#define OPENSIM_THROW(EXCEPTION, ...)                                \
+    throw EXCEPTION{__FILE__, __LINE__, __func__, __VA_ARGS__};
+
+#define OPENSIM_THROW_IF(CONDITION, EXCEPTION, ...)                  \
+    {                                                                \
+    if(CONDITION)                                                    \
+        OPENSIM_THROW(EXCEPTION, __VA_ARGS__)                        \
+    }
+
+
 
 /**
  * A class for basic exception functionality.
+ * To create exception classes in OpenSim, use the following guidlines.
+ * If the intention is the derive from an exception named (say) BaseException
+ * that is part of OpenSim, use the following blueprint:
+ * \code{.cpp}
+ *     class MyNewException : public BaseException {
+ *     public:
+ *         MyNewException(const std::string& file,
+ *                        size_t line,
+ *                        const std::string& func,
+ *                        <more parameters as appropriate>) :
+ *         BaseException(flie, line, func) {
+ *             std::string message = <create the desired message>;
+ *             addMessage(message);
+ *         }
+ *     };
+ * \endcode
+ * Exception class manages the concatenation of error messages from all the 
+ * derived classes. When creating new exceptions, remember to call addMessage()
+ * as shown above if the exception class does have any error message.
```

Comment:
Yes. It appears in doxygen. I just made another commit to have it only for developer doxygen.